helix-core/src/main/java/org/apache/helix/task/ThreadCountBasedTaskAssignmentCalculator.java

```diff
@@ -65,10 +66,26 @@ public ThreadCountBasedTaskAssignmentCalculator(TaskAssigner taskAssigner,
       Map<String, IdealState> idealStateMap) {
     Map<String, TaskConfig> taskMap = jobCfg.getTaskConfigMap();
     Map<String, Integer> taskIdMap = jobCtx.getTaskIdPartitionMap();
+    // Check if a gap exists in the context due to previouslys
+    // removed tasks. If yes, the missing pIDs should be considered
+    // for newly added tasks
+    Set<Integer> existingPartitions = jobCtx.getPartitionSet();
+    Set<Integer> missingPartitions = new HashSet<>();
+    if (existingPartitions.size() != 0) {
+      for (int pId = 0; pId < Collections.max(existingPartitions); pId++) {
+        if (!existingPartitions.contains(pId)) {
+          missingPartitions.add(pId);
+        }
+      }
+    }
     for (TaskConfig taskCfg : taskMap.values()) {
       String taskId = taskCfg.getId();
-      int nextPartition = jobCtx.getPartitionSet().size();
       if (!taskIdMap.containsKey(taskId)) {
+        int nextPartition = jobCtx.getPartitionSet().size();
+        if (missingPartitions.size()!=0) {
+          nextPartition = missingPartitions.iterator().next();
+          missingPartitions.remove(nextPartition);
+        }
         jobCtx.setTaskIdForPartition(nextPartition, taskId);
       }
     }
```

Comment:
Is this function exactly the same as the previous one? Is there a way we can combine them and not repeat?

Reply:
I thought about it actually. It was already duplicated. I think we keep a generic one for backward compatibility.